hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/common/SparkReaderContextFactory.java

```diff
@@ -88,6 +90,9 @@ class SparkReaderContextFactory implements ReaderContextFactory<InternalRow> {
     // Spark parquet reader has to be instantiated on the driver and broadcast to the executors
     SparkParquetReader parquetFileReader = sparkAdapter.createParquetFileReader(false, sqlConf, options, configs);
     parquetReaderBroadcast = jsc.broadcast(parquetFileReader);
+    // Broadcast: TableConfig.
+    HoodieTableConfig tableConfig = metaClient.getTableConfig();
```

Comment:
The `HoodieFileGroupReader.props` already contains all the write-config options  plus the table-config options, we can utilize that once the key gen related logic been moved to the `HoodieFileGroupReader`.

Also we already have the broadcasted `SerializableConfiguration`, can we just set all the table options there and reuse it, just like what we do in `HoodieFileGroupReader`: one props to cover all the config options.

Reply:
The `SerializableConfiguration` is just the hadoop and IO focused props. The table config is more focused on the configuration of the table so I don't think it is wise to mix these. It is better to be intentional here.

The other option is to have the factory take in a `HoodieTableConfig` or properties. This is what I originally had but it was recommended to do it this way by @yihua

Reply:
As Tim mentioned `SerializableConfiguration` wraps the hadoop configs shared in a Spark session which do not contain Hudi configs and it's unsafe to add table-specific configs there.  And given the table config object is small it's OK to have it this way now, and let's remove the table config from the reader context after subsequent refactoring once the Hudi semantics are removed from the reader context class.  @danny0405 wdyt

Reply:
> The SerializableConfiguration is just the hadoop and IO focused props.

Not really, the `SparkReaderContextFactory#getHadoopConfiguration` sets up option like schema evolution and parquet.

I kind of think the `ReaderContext` is kind of developer API, should we expose component directly instead of Hudi related configs in case the developer has no good insight of the numerous Hudi options.

Reply:
The schema evolution should not be in the hadoop configuration to begin with in my opinion. It is not related to spark or hadoop, it is part of the Hudi logic therefore should be in some configuration for the Hudi reader.

It also should not be overriding the user provided configuration like it does today. It limits the user's ability to customize their job and read their data after upgrading. 

I think we generally will need to rethink this approach but is it a blocker for this PR?

Reply:
maybe, but the write config plus table config params in reader context may be very confusing to developers.

Reply:
The HoodieWriteConfig is not included in the reader context, what do you mean by write config in this context?

Reply:
many write config are shared with the reader too, what I meant here is do we want to expose multiple configurations to the reader context for the final version?

Reply:
For the readers in general, we can only rely on the HoodieTableConfig and the timeline for the information we need for reading. We cannot use any write configs because the reader may have no knowledge of the writer.